jaseci_kit/jaseci_kit/encoders/utils/models.py

```diff
@@ -176,16 +176,53 @@ def __init__(self, config, *inputs, **kwargs):
         else:
             self.cont_bert = kwargs['cont_bert']
             self.cand_bert = kwargs['cand_bert']
+        self.input_size = None
+        self.l_type = kwargs['loss_type']
+        self.l_func = kwargs['loss_function']
+
+    def loss_func(self, candidate_vec, context_vec, labels, l_type='cos',
+                  l_func='MSE'):
+        candidate_vec = candidate_vec.squeeze(1)
+        labels = labels.squeeze(1)
+        if l_type == 'dot':
+            dot_product = torch.matmul(
+                context_vec, candidate_vec.t())
+            mask = torch.eye(self.input_size).to(
+                candidate_vec.device)
+            dot_product = 1 - F.log_softmax(dot_product, dim=-1) * mask
+            if l_func == 'contrastive':
+                dot_product = dot_product.mean(dim=-1)
+                loss = 0.5 * (labels.float() * dot_product.pow(2) +
+                              (1 - labels).float() *
+                              F.relu(0.5 - dot_product).pow(2))
+                loss = loss.mean()
+            else:
+                loss_fnct = nn.MSELoss()
+                loss = loss_fnct(dot_product.mean(dim=-1), target=labels)
+        else:
+            cos = nn.CosineSimilarity()
+            if l_func == 'contrastive':
+                loss_angle = 1 - cos(context_vec, candidate_vec)
+                loss = 0.5 * (labels.float() * loss_angle.pow(2) +
+                              (1 - labels).float() *
+                              F.relu(0.5 - loss_angle).pow(2))
+                loss = loss.mean(dim=-1)
+            else:
+                loss_angle = cos(context_vec, candidate_vec)
+                loss_fnct = nn.MSELoss()
+                loss = loss_fnct(loss_angle, target=labels)
+        return loss
 
     def forward(self, context_input_ids=None, context_input_masks=None,
                 candidate_input_ids=None, candidate_input_masks=None,
                 labels=None, get_embedding=None, mode="train", pooling="mean"):
         # only select the first candidate (whose lbl==1)
-        if labels is not None:
-            candidate_input_ids = candidate_input_ids[:, 0, :].unsqueeze(1)
-            candidate_input_masks = candidate_input_masks[:, 0, :].unsqueeze(1)
+        # if labels is not None:
```

Comment:
remove unncessary code please.

Reply:
this code was used when, the positive and negative sample for 1 datapoint is combined